metric-publishers/emf-metric-publisher/src/main/java/software/amazon/awssdk/metrics/publishers/emf/MetricEmfConverter.java

```diff
@@ -0,0 +1,368 @@
+            }
+        }
+
+        return createEmfStrings(aggregatedMetrics);
+    }
+
+    private List<String> createEmfStrings(Map<String, List<Object>> aggregatedMetrics) {
+        List<String> emfStrings = new ArrayList<>();
+        Map<String, List<Object>> currentMetricBatch = new HashMap<>();
+        Set<String> currentMetricNames = new HashSet<>();
+
+        // Process metric names and their values
+        for (Map.Entry<String, List<Object>> entry : aggregatedMetrics.entrySet()) {
+            String metricName = entry.getKey();
+            List<Object> values = entry.getValue();
+
+            // Drop large value arrays into chunks of 100
+            if (values.size() > 100) {
+                values = values.subList(0, 100);
+                logger.warn(() -> "Some AWS SDK client-side metric data have been dropped because it exceeds the cloudwatch "
+                                 + "requirements.");
+            }
+            // If adding this metric would exceed 100 metrics, create new batch
+            if (currentMetricNames.size() >= 100) {
+                emfStrings.add(createEmfString(currentMetricBatch, currentMetricNames));
+                currentMetricBatch = new HashMap<>();
+                currentMetricNames = new HashSet<>();
+            }
+
+            currentMetricBatch.put(metricName, values);
+            if (!(values.get(0) instanceof String)) {
+                currentMetricNames.add(metricName);
+            }
+
+        }
+
+        emfStrings.add(createEmfString(currentMetricBatch, currentMetricNames));
+
+        return emfStrings;
+    }
+
+
+    private String createEmfString(Map<String, List<Object>> metrics, Set<String> metricNames) {
+
+        try {
+            JsonWriter jsonWriter = JsonWriter.create();
+            jsonWriter.writeStartObject();
+
+            writeAwsObject(jsonWriter, metricNames);
+            // Write metric values
+            writeMetricValues(jsonWriter, metrics);
+
+            jsonWriter.writeEndObject(); // End root object
+
+            return new String(jsonWriter.getBytes(), StandardCharsets.UTF_8);
+
+        } catch (SdkClientException e) {
+            logger.error(() -> "Failed to create EMF format string");
+            throw e;
+        }
+    }
+
+    private void writeAwsObject(JsonWriter jsonWriter, Set<String> metricNames) {
+        // Start with _aws section
+        jsonWriter.writeFieldName("_aws");
+        jsonWriter.writeStartObject();
+
+        // Write Timestamp
+        jsonWriter.writeFieldName("Timestamp");
+        //jsonWriter.writeValue(Instant.now().toEpochMilli());
+        //Unit Test
+        jsonWriter.writeValue(12345678);
+
+
+        jsonWriter.writeFieldName("LogGroupName");
+        jsonWriter.writeValue(config.getLogGroupName());
+
+        // Write CloudWatchMetrics array
+        jsonWriter.writeFieldName("CloudWatchMetrics");
+        jsonWriter.writeStartArray();
+        jsonWriter.writeStartObject();
```

Comment:
Is this correct? It seems this CloudWatchMetrics only has one object

Reply:
I think so. In the example https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Embedded_Metric_Format_Specification.html, CloudWatchMetrics should have one object, because we only have one namespace and one set of dimensions